File: src/main/scala/ff/compiler/Token.scala
package ff.compiler
import ff.compiler.Syntax_._

import ff.core.Array_._

import ff.core.ArrayBuilder_._

import ff.core.Bool_._

import ff.core.Char_._

import ff.core.Core_._

import ff.core.FileSystem_._

import ff.core.Int_._

import ff.core.List_._

import ff.core.Log_._

import ff.core.Map_._

import ff.core.Nothing_._

import ff.core.Option_._

import ff.core.Pair_._

import ff.core.Set_._

import ff.core.String_._

import ff.core.System_._

import ff.core.Try_._

import ff.core.Unit_._
object Token_ {

case class Token(file_ : ff.core.String_.String, code_ : ff.core.String_.String, kind_ : ff.compiler.Token_.TokenKind, startLine_ : ff.core.Int_.Int, startLineOffset_ : ff.core.Int_.Int, startOffset_ : ff.core.Int_.Int, stopLine_ : ff.core.Int_.Int, stopLineOffset_ : ff.core.Int_.Int, stopOffset_ : ff.core.Int_.Int)

sealed abstract class TokenKind extends Product with Serializable
case class LEnd() extends TokenKind
case class LString() extends TokenKind
case class LChar() extends TokenKind
case class LInt() extends TokenKind
case class LFloat() extends TokenKind
case class LKeyword() extends TokenKind
case class LNamespace() extends TokenKind
case class LLower() extends TokenKind
case class LUpper() extends TokenKind
case class LWildcard() extends TokenKind
case class LBracketLeft() extends TokenKind
case class LBracketRight() extends TokenKind
case class LOperator() extends TokenKind
case class LComma() extends TokenKind
case class LSeparator() extends TokenKind
case class LDot() extends TokenKind
case class LSemicolon() extends TokenKind
case class LPipe() extends TokenKind
case class LColon() extends TokenKind
case class LDotDotDot() extends TokenKind
case class LArrowThick() extends TokenKind
case class LAssign() extends TokenKind
case class LAssignPlus() extends TokenKind
case class LAssignMinus() extends TokenKind
case class LAssignLink() extends TokenKind


def Token_at(token_ : ff.compiler.Token_.Token) : ff.compiler.Syntax_.Location = (token_) match {
case (token_) =>
ff.compiler.Syntax_.Location(file_ = token_.file_, line_ = token_.startLine_, column_ = ((token_.startOffset_ - token_.startLineOffset_) + 1))
}

def Token_raw(token_ : ff.compiler.Token_.Token) : ff.core.String_.String = (token_) match {
case (token_) =>
ff.core.String_.String_slice(self_ = token_.code_, from_ = token_.startOffset_, until_ = token_.stopOffset_)
}

def Token_is(token_ : ff.compiler.Token_.Token, kind1_ : ff.compiler.Token_.TokenKind) : ff.core.Bool_.Bool = (token_, kind1_) match {
case (token_, _) =>
(token_.kind_ == kind1_)
}

def Token_is2(token_ : ff.compiler.Token_.Token, kind1_ : ff.compiler.Token_.TokenKind, kind2_ : ff.compiler.Token_.TokenKind) : ff.core.Bool_.Bool = (token_, kind1_, kind2_) match {
case (token_, _, _) =>
((token_.kind_ == kind1_) || (token_.kind_ == kind2_))
}

def Token_is3(token_ : ff.compiler.Token_.Token, kind1_ : ff.compiler.Token_.TokenKind, kind2_ : ff.compiler.Token_.TokenKind, kind3_ : ff.compiler.Token_.TokenKind) : ff.core.Bool_.Bool = (token_, kind1_, kind2_, kind3_) match {
case (token_, _, _, _) =>
(((token_.kind_ == kind1_) || (token_.kind_ == kind2_)) || (token_.kind_ == kind3_))
}

def Token_rawIs(token_ : ff.compiler.Token_.Token, value_ : ff.core.String_.String) : ff.core.Bool_.Bool = (token_, value_) match {
case (token_, _) =>
(((token_.stopOffset_ - token_.startOffset_) == ff.core.String_.String_size(self_ = value_)) && ff.core.String_.String_startsWith(self_ = token_.code_, prefix_ = value_, offset_ = token_.startOffset_))
}

def TokenKind_beforeSeparator(self_ : ff.compiler.Token_.TokenKind) : ff.core.Bool_.Bool = (self_) match {
case (self_) =>
pipe_dot(self_)({
case (ff.compiler.Token_.LEnd()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LString()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LChar()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LInt()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LFloat()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LKeyword()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LNamespace()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LLower()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LUpper()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LWildcard()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LBracketLeft()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LBracketRight()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LOperator()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LComma()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LSeparator()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LDot()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LSemicolon()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LPipe()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LColon()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LDotDotDot()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LArrowThick()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssign()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssignPlus()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssignMinus()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssignLink()) =>
ff.core.Bool_.False()
})
}

def TokenKind_afterSeparator(self_ : ff.compiler.Token_.TokenKind) : ff.core.Bool_.Bool = (self_) match {
case (self_) =>
pipe_dot(self_)({
case (ff.compiler.Token_.LEnd()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LString()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LChar()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LInt()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LFloat()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LKeyword()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LNamespace()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LLower()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LUpper()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LWildcard()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LBracketLeft()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LBracketRight()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LOperator()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LComma()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LSeparator()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LDot()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LSemicolon()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LPipe()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LColon()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LDotDotDot()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LArrowThick()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssign()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssignPlus()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssignMinus()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssignLink()) =>
ff.core.Bool_.False()
})
}

def TokenKind_afterKeyword(self_ : ff.compiler.Token_.TokenKind) : ff.core.Bool_.Bool = (self_) match {
case (self_) =>
pipe_dot(self_)({
case (ff.compiler.Token_.LEnd()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LString()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LChar()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LInt()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LFloat()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LKeyword()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LNamespace()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LLower()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LUpper()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LWildcard()) =>
ff.core.Bool_.True()
case (ff.compiler.Token_.LBracketLeft()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LBracketRight()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LOperator()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LComma()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LSeparator()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LDot()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LSemicolon()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LPipe()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LColon()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LDotDotDot()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LArrowThick()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssign()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssignPlus()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssignMinus()) =>
ff.core.Bool_.False()
case (ff.compiler.Token_.LAssignLink()) =>
ff.core.Bool_.False()
})
}


}

File: src/main/js/ff/compiler/Token.mjs
import * as ff_compiler_Token from "../../ff/compiler/Token.mjs"

import * as ff_compiler_Syntax from "../../ff/compiler/Syntax.mjs"

import * as ff_core_Array from "../../ff/core/Array.mjs"

import * as ff_core_ArrayBuilder from "../../ff/core/ArrayBuilder.mjs"

import * as ff_core_Bool from "../../ff/core/Bool.mjs"

import * as ff_core_Char from "../../ff/core/Char.mjs"

import * as ff_core_Core from "../../ff/core/Core.mjs"

import * as ff_core_Duration from "../../ff/core/Duration.mjs"

import * as ff_core_FileSystem from "../../ff/core/FileSystem.mjs"

import * as ff_core_Float from "../../ff/core/Float.mjs"

import * as ff_core_Int from "../../ff/core/Int.mjs"

import * as ff_core_List from "../../ff/core/List.mjs"

import * as ff_core_Log from "../../ff/core/Log.mjs"

import * as ff_core_Map from "../../ff/core/Map.mjs"

import * as ff_core_Nothing from "../../ff/core/Nothing.mjs"

import * as ff_core_Option from "../../ff/core/Option.mjs"

import * as ff_core_Pair from "../../ff/core/Pair.mjs"

import * as ff_core_Set from "../../ff/core/Set.mjs"

import * as ff_core_String from "../../ff/core/String.mjs"

import * as ff_core_System from "../../ff/core/System.mjs"

import * as ff_core_Try from "../../ff/core/Try.mjs"

import * as ff_core_Unit from "../../ff/core/Unit.mjs"

// type Token
export function Token(file_, code_, kind_, startLine_, startLineOffset_, startOffset_, stopLine_, stopLineOffset_, stopOffset_) {
return {file_, code_, kind_, startLine_, startLineOffset_, startOffset_, stopLine_, stopLineOffset_, stopOffset_};
}

// type TokenKind
const LEnd$ = {LEnd: true};
export function LEnd() {
return LEnd$;
}
const LString$ = {LString: true};
export function LString() {
return LString$;
}
const LChar$ = {LChar: true};
export function LChar() {
return LChar$;
}
const LInt$ = {LInt: true};
export function LInt() {
return LInt$;
}
const LFloat$ = {LFloat: true};
export function LFloat() {
return LFloat$;
}
const LKeyword$ = {LKeyword: true};
export function LKeyword() {
return LKeyword$;
}
const LNamespace$ = {LNamespace: true};
export function LNamespace() {
return LNamespace$;
}
const LLower$ = {LLower: true};
export function LLower() {
return LLower$;
}
const LUpper$ = {LUpper: true};
export function LUpper() {
return LUpper$;
}
const LWildcard$ = {LWildcard: true};
export function LWildcard() {
return LWildcard$;
}
const LBracketLeft$ = {LBracketLeft: true};
export function LBracketLeft() {
return LBracketLeft$;
}
const LBracketRight$ = {LBracketRight: true};
export function LBracketRight() {
return LBracketRight$;
}
const LOperator$ = {LOperator: true};
export function LOperator() {
return LOperator$;
}
const LComma$ = {LComma: true};
export function LComma() {
return LComma$;
}
const LSeparator$ = {LSeparator: true};
export function LSeparator() {
return LSeparator$;
}
const LDot$ = {LDot: true};
export function LDot() {
return LDot$;
}
const LSemicolon$ = {LSemicolon: true};
export function LSemicolon() {
return LSemicolon$;
}
const LPipe$ = {LPipe: true};
export function LPipe() {
return LPipe$;
}
const LColon$ = {LColon: true};
export function LColon() {
return LColon$;
}
const LDotDotDot$ = {LDotDotDot: true};
export function LDotDotDot() {
return LDotDotDot$;
}
const LArrowThick$ = {LArrowThick: true};
export function LArrowThick() {
return LArrowThick$;
}
const LAssign$ = {LAssign: true};
export function LAssign() {
return LAssign$;
}
const LAssignPlus$ = {LAssignPlus: true};
export function LAssignPlus() {
return LAssignPlus$;
}
const LAssignMinus$ = {LAssignMinus: true};
export function LAssignMinus() {
return LAssignMinus$;
}
const LAssignLink$ = {LAssignLink: true};
export function LAssignLink() {
return LAssignLink$;
}





export function Token_at(token_) {
return ff_compiler_Syntax.Location(token_.file_, token_.startLine_, ((token_.startOffset_ - token_.startLineOffset_) + 1))
}

export function Token_raw(token_) {
return ff_core_String.String_slice(token_.code_, token_.startOffset_, token_.stopOffset_)
}

export function Token_is(token_, kind1_) {
return (token_.kind_ == kind1_)
}

export function Token_is2(token_, kind1_, kind2_) {
return ((token_.kind_ == kind1_) || (token_.kind_ == kind2_))
}

export function Token_is3(token_, kind1_, kind2_, kind3_) {
return (((token_.kind_ == kind1_) || (token_.kind_ == kind2_)) || (token_.kind_ == kind3_))
}

export function Token_rawIs(token_, value_) {
return (((token_.stopOffset_ - token_.startOffset_) == ff_core_String.String_size(value_)) && ff_core_String.String_startsWith(token_.code_, value_, token_.startOffset_))
}

export function TokenKind_beforeSeparator(self_) {
{
const _1 = self_
{
if(_1.LEnd) {
return false
return
}
}
{
if(_1.LString) {
return true
return
}
}
{
if(_1.LChar) {
return true
return
}
}
{
if(_1.LInt) {
return true
return
}
}
{
if(_1.LFloat) {
return true
return
}
}
{
if(_1.LKeyword) {
return true
return
}
}
{
if(_1.LNamespace) {
return false
return
}
}
{
if(_1.LLower) {
return true
return
}
}
{
if(_1.LUpper) {
return true
return
}
}
{
if(_1.LWildcard) {
return true
return
}
}
{
if(_1.LBracketLeft) {
return false
return
}
}
{
if(_1.LBracketRight) {
return true
return
}
}
{
if(_1.LOperator) {
return false
return
}
}
{
if(_1.LComma) {
return false
return
}
}
{
if(_1.LSeparator) {
return false
return
}
}
{
if(_1.LDot) {
return false
return
}
}
{
if(_1.LSemicolon) {
return false
return
}
}
{
if(_1.LPipe) {
return false
return
}
}
{
if(_1.LColon) {
return false
return
}
}
{
if(_1.LDotDotDot) {
return false
return
}
}
{
if(_1.LArrowThick) {
return false
return
}
}
{
if(_1.LAssign) {
return false
return
}
}
{
if(_1.LAssignPlus) {
return false
return
}
}
{
if(_1.LAssignMinus) {
return false
return
}
}
{
if(_1.LAssignLink) {
return false
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function TokenKind_afterSeparator(self_) {
{
const _1 = self_
{
if(_1.LEnd) {
return false
return
}
}
{
if(_1.LString) {
return true
return
}
}
{
if(_1.LChar) {
return true
return
}
}
{
if(_1.LInt) {
return true
return
}
}
{
if(_1.LFloat) {
return true
return
}
}
{
if(_1.LKeyword) {
return true
return
}
}
{
if(_1.LNamespace) {
return true
return
}
}
{
if(_1.LLower) {
return true
return
}
}
{
if(_1.LUpper) {
return true
return
}
}
{
if(_1.LWildcard) {
return true
return
}
}
{
if(_1.LBracketLeft) {
return true
return
}
}
{
if(_1.LBracketRight) {
return false
return
}
}
{
if(_1.LOperator) {
return false
return
}
}
{
if(_1.LComma) {
return false
return
}
}
{
if(_1.LSeparator) {
return false
return
}
}
{
if(_1.LDot) {
return false
return
}
}
{
if(_1.LSemicolon) {
return false
return
}
}
{
if(_1.LPipe) {
return false
return
}
}
{
if(_1.LColon) {
return false
return
}
}
{
if(_1.LDotDotDot) {
return true
return
}
}
{
if(_1.LArrowThick) {
return false
return
}
}
{
if(_1.LAssign) {
return false
return
}
}
{
if(_1.LAssignPlus) {
return false
return
}
}
{
if(_1.LAssignMinus) {
return false
return
}
}
{
if(_1.LAssignLink) {
return false
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function TokenKind_afterKeyword(self_) {
{
const _1 = self_
{
if(_1.LEnd) {
return false
return
}
}
{
if(_1.LString) {
return true
return
}
}
{
if(_1.LChar) {
return true
return
}
}
{
if(_1.LInt) {
return true
return
}
}
{
if(_1.LFloat) {
return true
return
}
}
{
if(_1.LKeyword) {
return true
return
}
}
{
if(_1.LNamespace) {
return true
return
}
}
{
if(_1.LLower) {
return true
return
}
}
{
if(_1.LUpper) {
return true
return
}
}
{
if(_1.LWildcard) {
return true
return
}
}
{
if(_1.LBracketLeft) {
return false
return
}
}
{
if(_1.LBracketRight) {
return false
return
}
}
{
if(_1.LOperator) {
return false
return
}
}
{
if(_1.LComma) {
return false
return
}
}
{
if(_1.LSeparator) {
return false
return
}
}
{
if(_1.LDot) {
return false
return
}
}
{
if(_1.LSemicolon) {
return false
return
}
}
{
if(_1.LPipe) {
return false
return
}
}
{
if(_1.LColon) {
return false
return
}
}
{
if(_1.LDotDotDot) {
return false
return
}
}
{
if(_1.LArrowThick) {
return false
return
}
}
{
if(_1.LAssign) {
return false
return
}
}
{
if(_1.LAssignPlus) {
return false
return
}
}
{
if(_1.LAssignMinus) {
return false
return
}
}
{
if(_1.LAssignLink) {
return false
return
}
}
throw new Error('Unexhaustive pattern match')
}
}


